### PR TITLE
[P3-A3-WP3] Replace hardcoded type=system/session_id polling with StreamParser.parse_line()

### PR DESCRIPTION
**Pull request by Trecek**
## Summary

Replace the inline `fast_loads` + `type=="system"` JSON parsing in `_extract_stdout_session_id` with `StreamParser.parse_line()` calls, falling back to the current logic when no parser is provided. Thread the `stream_parser` parameter from `run_managed_async` through the anyio task group down to `_extract_stdout_session_id`. Update the `SubprocessRunner` protocol and all its concrete implementations (`DefaultSubprocessRunner`, `RecordingSubprocessRunner`, `ReplayingSubprocessRunner`) to maintain the documented signature-mirroring invariant. Update contract tests that hardcode parameter sets.

## Goal

Replace hardcoded type=system/session_id polling in _extract_stdout_session_id with StreamParser.parse_line() calls, falling back to current logic when parser is None.

## Context

- Phase: CodingAgentBackend Protocol and ClaudeCodeBackend Extraction (Milestone: 3-codingagentbackend-protocol-and-claudecodebackend-extraction)
- Assignment: P3-A3 (Refactor race engine to use StreamParser)
- Depends on: P3-A1-WP1, P3-A3-WP2
- Depended on by: None

## Deliverables

- `Updated _extract_stdout_session_id with StreamParser support`
- `Updated run_managed_async threading stream_parser`
- `Tests for both paths`

## Technical Steps

1. Add stream_parser: StreamParser | None = None to _extract_stdout_session_id signature
2. Add branch: if stream_parser provided, call parse_line(line) and read event.session_id
3. Retain existing fast_loads + type==system logic as fallback when stream_parser is None
4. Thread stream_parser from run_managed_async to _extract_stdout_session_id in task group
5. Update imports with TYPE_CHECKING guard for StreamParser and SessionEvent
6. Add tests for both StreamParser path and None fallback path

## Acceptance Criteria

- With stream_parser: uses parse_line, reads session_id without fast_loads
- Without stream_parser: identical to pre-refactor behavior
- run_managed_async passes stream_parser to _extract_stdout_session_id
- Existing tests pass
- New tests for both branches added

### Modified (●):

- `● src/autoskillit/core/types/_type_subprocess.py`
- `● src/autoskillit/execution/process/__init__.py`
- `● src/autoskillit/execution/process/_process_race.py`
- `● src/autoskillit/execution/recording.py`
- `● tests/contracts/test_protocol_satisfaction.py`
- `● tests/execution/test_process_race.py`

Closes #2484

## Implementation Plan

Plan file: `/home/talon/projects/autoskillit-runs/impl-20260518-234141-687008/.autoskillit/temp/make-plan/p3_a3_wp3_stream_parser_session_id_plan_2026-05-18_234500.md`

🤖 Generated with [Claude Code](https://claude.com/claude-code) via AutoSkillit

## Token Usage Summary

| Step | Model | count | uncached | output | cache_read | peak_ctx | turns | cache_write | time |
|------|-------|-------|----------|--------|------------|----------|-------|-------------|------|
| plan | claude-sonnet-4-6 | 1 | 2.2k | 27.6k | 1.4M | 104.3k | 271 | 94.8k | 12m 24s |
| verify | claude-opus-4-6 | 1 | 35 | 8.9k | 1.3M | 98.1k | 112 | 96.5k | 5m 0s |
| implement* | MiniMax-M2.7-highspeed | 1 | 1.7M | 13.1k | 1.5M | 70.3k | 139 | 82.0k | 6m 15s |
| prepare_pr* | MiniMax-M2.7 | 1 | 66.1k | 3.3k | 178.1k | 29.2k | 19 | 42.0k | 2m 16s |
| compose_pr* | MiniMax-M2.7 | 1 | 39.2k | 2.1k | 310.7k | 0 | 24 | 0 | 1m 15s |
| review_pr | claude-opus-4-6 | 2 | 67 | 51.8k | 1.1M | 70.2k | 71 | 113.1k | 11m 7s |
| resolve_review | claude-opus-4-6 | 1 | 56 | 12.8k | 1.0M | 61.1k | 69 | 46.9k | 7m 21s |
| **Total** | | | 1.8M | 119.7k | 6.9M | 104.3k | | 475.3k | 45m 42s |

\* *Step used a non-Anthropic provider; caching behavior may differ.*

## Token Efficiency

| Step | LoC Changed | cache_read/LoC | cache_write/LoC | output/LoC |
|------|-------------|----------------|-----------------|------------|
| plan | 0 | — | — | — |
| verify | 0 | — | — | — |
| implement | 172 | 8886.3 | 476.8 | 76.4 |
| prepare_pr | 0 | — | — | — |
| compose_pr | 0 | — | — | — |
| review_pr | 0 | — | — | — |
| resolve_review | 10 | 102292.1 | 4688.8 | 1277.5 |
| **Total** | **182** | 37744.1 | 2611.6 | 657.8 |

## Model Usage Breakdown

| Model | steps | uncached | output | cache_read | cache_write | time |
|-------|-------|----------|--------|------------|-------------|------|
| claude-sonnet-4-6 | 1 | 2.2k | 27.6k | 1.4M | 94.8k | 12m 24s |
| claude-opus-4-6 | 3 | 158 | 73.5k | 3.4M | 256.5k | 23m 29s |
| MiniMax-M2.7-highspeed | 1 | 1.7M | 13.1k | 1.5M | 82.0k | 6m 15s |
| MiniMax-M2.7 | 2 | 105.2k | 5.4k | 488.8k | 42.0k | 3m 32s |

### CODE REVIEW

File: src/autoskillit/core/types/_type_subprocess.py [hunk 1]
```diff
@@ -162,4 +162,5 @@ def __call__(
         max_extension_seconds: float = 7200,
         marker_dir: Path | None = None,
         session_id: str | None = None,
+        stream_parser: Any | None = None,
```
Comment (Trecek) [inline src/autoskillit/core/types/_type_subprocess.py:165]: [warning] cohesion: Protocol signature uses `Any | None` for `stream_parser` while all concrete implementations (process/__init__.py, recording.py, _process_race.py) use `StreamParser | None`. The protocol should use the concrete type to provide type safety and remain consistent with its implementations.

**Review by Trecek — COMMENTED** [review]
AutoSkillit PR Review — Verdict: approved_with_comments
